Building against 6.4+ kernel fails

Building on an Ubuntu 22.04 system, on 6.3.4 it builds without issue, on 6.4.x it fails with the following. Same problem on 6.4.3, 6.4.2 and 6.4.1.

> DKMS make.log for gasket-1.0 for kernel 6.4.1-060401-generic (x86_64)
> Thu Jul 13 07:21:12 PM AWST 2023
> make: Entering directory '/usr/src/linux-headers-6.4.1-060401-generic'
> warning: the compiler differs from the one used to build the kernel
>   The kernel was built by: x86_64-linux-gnu-gcc-12 (Ubuntu 12.3.0-1ubuntu1) 12.3.0
>   You are using:           gcc-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_core.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_ioctl.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_interrupt.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_page_table.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_sysfs.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/apex_driver.o
> /var/lib/dkms/gasket/1.0/build/gasket_core.c: In function ‘gasket_register_device’:
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:41: error: passing argument 1 of ‘class_create’ from incompatible pointer type [-Werror=incompatible-pointer-types]
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                              ~~~~~~~~~~~^~~~~~~~
>       |                                         |
>       |                                         struct module *
> In file included from ./include/linux/device.h:31,
>                  from ./include/linux/cdev.h:8,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.h:11,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.c:12:
> ./include/linux/device/class.h:230:54: note: expected ‘const char *’ but argument is of type ‘struct module *’
>   230 | struct class * __must_check class_create(const char *name);
>       |                                          ~~~~~~~~~~~~^~~~
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:17: error: too many arguments to function ‘class_create’
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                 ^~~~~~~~~~~~
> ./include/linux/device/class.h:230:29: note: declared here
>   230 | struct class * __must_check class_create(const char *name);
>       |                             ^~~~~~~~~~~~
> cc1: some warnings being treated as errors
> make[1]: *** [scripts/Makefile.build:252: /var/lib/dkms/gasket/1.0/build/gasket_core.o] Error 1
> make[1]: *** Waiting for unfinished jobs....
> make: *** [Makefile:2032: /var/lib/dkms/gasket/1.0/build] Error 2
> make: Leaving directory '/usr/src/linux-headers-6.4.1-060401-generic'

Tested again on a fresh Ubuntu 22.04 install and same results on 6.4.3

> DKMS make.log for gasket-1.0 for kernel 6.4.3-060403-generic (x86_64)
> Thu Jul 13 01:12:18 PM UTC 2023
> make: Entering directory '/usr/src/linux-headers-6.4.3-060403-generic'
> warning: the compiler differs from the one used to build the kernel
>   The kernel was built by: x86_64-linux-gnu-gcc-12 (Ubuntu 12.3.0-1ubuntu1) 12.3.0
>   You are using:           gcc-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_core.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_ioctl.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_interrupt.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_page_table.o
> /var/lib/dkms/gasket/1.0/build/gasket_core.c: In function ‘gasket_register_device’:
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:41: error: passing argument 1 of ‘class_create’ from incompatible pointer type [-Werror=incompatible-pointer-types]
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                              ~~~~~~~~~~~^~~~~~~~
>       |                                         |
>       |                                         struct module *
> In file included from ./include/linux/device.h:31,
>                  from ./include/linux/cdev.h:8,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.h:11,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.c:12:
> ./include/linux/device/class.h:230:54: note: expected ‘const char *’ but argument is of type ‘struct module *’
>   230 | struct class * __must_check class_create(const char *name);
>       |                                          ~~~~~~~~~~~~^~~~
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:17: error: too many arguments to function ‘class_create’
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                 ^~~~~~~~~~~~
> ./include/linux/device/class.h:230:29: note: declared here
>   230 | struct class * __must_check class_create(const char *name);
>       |                             ^~~~~~~~~~~~
> cc1: some warnings being treated as errors
> make[1]: *** [scripts/Makefile.build:252: /var/lib/dkms/gasket/1.0/build/gasket_core.o] Error 1
> make[1]: *** Waiting for unfinished jobs....
> make: *** [Makefile:2032: /var/lib/dkms/gasket/1.0/build] Error 2
> make: Leaving directory '/usr/src/linux-headers-6.4.3-060403-generic'

And testing on 23.04 same mainline 6.4.3 kernel

> DKMS make.log for gasket-1.0 for kernel 6.4.3-060403-generic (x86_64)
> Thu Jul 13 01:12:11 PM UTC 2023
> make: Entering directory '/usr/src/linux-headers-6.4.3-060403-generic'
> warning: the compiler differs from the one used to build the kernel
>   The kernel was built by: x86_64-linux-gnu-gcc-12 (Ubuntu 12.3.0-1ubuntu1) 12.3.0
>   You are using:           gcc-12 (Ubuntu 12.2.0-17ubuntu1) 12.2.0
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_core.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_ioctl.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_page_table.o
>   CC [M]  /var/lib/dkms/gasket/1.0/build/gasket_interrupt.o
> /var/lib/dkms/gasket/1.0/build/gasket_core.c: In function ‘gasket_register_device’:
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:41: error: passing argument 1 of ‘class_create’ from incompatible pointer type [-Werror=incompatible-pointer-types]
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                              ~~~~~~~~~~~^~~~~~~~
>       |                                         |
>       |                                         struct module *
> In file included from ./include/linux/device.h:31,
>                  from ./include/linux/cdev.h:8,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.h:11,
>                  from /var/lib/dkms/gasket/1.0/build/gasket_core.c:12:
> ./include/linux/device/class.h:230:54: note: expected ‘const char *’ but argument is of type ‘struct module *’
>   230 | struct class * __must_check class_create(const char *name);
>       |                                          ~~~~~~~~~~~~^~~~
> /var/lib/dkms/gasket/1.0/build/gasket_core.c:1841:17: error: too many arguments to function ‘class_create’
>  1841 |                 class_create(driver_desc->module, driver_desc->name);
>       |                 ^~~~~~~~~~~~
> ./include/linux/device/class.h:230:29: note: declared here
>   230 | struct class * __must_check class_create(const char *name);
>       |                             ^~~~~~~~~~~~
> cc1: some warnings being treated as errors
> make[1]: *** [scripts/Makefile.build:252: /var/lib/dkms/gasket/1.0/build/gasket_core.o] Error 1
> make[1]: *** Waiting for unfinished jobs....
> make: *** [Makefile:2032: /var/lib/dkms/gasket/1.0/build] Error 2
> make: Leaving directory '/usr/src/linux-headers-6.4.3-060403-generic'

--- a/src/gasket_core.c
+++ b/src/gasket_core.c
@@ -27,6 +27,7 @@
 #include <linux/platform_device.h>
 #include <linux/printk.h>
 #include <linux/sched.h>
+#include <linux/version.h>
 
 #ifdef GASKET_KERNEL_TRACE_SUPPORT
 #define CREATE_TRACE_POINTS
@@ -1837,8 +1838,15 @@ int gasket_register_device(const struct gasket_driver_desc *driver_desc)
 	internal = &g_descs[desc_idx];
 	mutex_init(&internal->mutex);
 	memset(internal->devs, 0, sizeof(struct gasket_dev *) * GASKET_DEV_MAX);
-	internal->class =
-		class_create(driver_desc->module, driver_desc->name);
+
+    /* Function signature for `class_create()` is changed in kernel >= 6.4.x
+     * to only accept a single argument.
+     * */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 4, 0)
+    internal->class = class_create(driver_desc->module, driver_desc->name);
+#else
+    internal->class = class_create(driver_desc->name);
+#endif
 
 	if (IS_ERR(internal->class)) {
 		pr_err("Cannot register %s class [ret=%ld]\n",

